### DISCUSSION TITLE
Does the altstack attribute require compiler support to be used?

### DISCUSSION THREAD

**Discussion by violin0847** [NVIDIA/open-gpu-kernel-modules#299] [Q&A] [answered]
The alternate stack, altstack, is used for QA_BUILD RM builds. But when I define `QA_BUILD` and `__use_altstack__`, 'altstack' attribute directive ignored when compiling, and the compiled kernel modules do not work. 

Does the `altstack` attribute require compiler support to be used?
Have you customized the compiler?

For example:
- macros `NV_ENTER_RM_RUNTIME`, `NV_EXIT_RM_RUNTIME`, in [src/nvidia/arch/nvalloc/unix/include/osapi.h](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/d8f3bcff924776518f1e63286537c3cf365289ac/src/nvidia/arch/nvalloc/unix/include/osapi.h)
- `#define NV_API_CALL __attribute__((altstack(0)))` in [src/common/sdk/nvidia/inc/nv-kernel-interface-api.h](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/d8f3bcff924776518f1e63286537c3cf365289ac/src/common/sdk/nvidia/inc/nv-kernel-interface-api.h)

**Comment by aritger** [comment, marked as answer]
Yes, altstack requires a custom compiler.  We don't use altstack in the open-gpu-kernel-module builds and are in general trying to phase out altstack.

And to be clear: I don't believe QA_BUILD RM builds _require_ altstack.  The code in osapi.h you mention just gets enabled if both of QA_BUILD and altstack are enabled.
